Initial draft of SDK Configuration specification (#14)
momentohq/standards-and-practices: docs/client-specifications/configuration.md

<img src="https://docs.momentohq.com/img/logo.svg" alt="logo" width="400"/>

# Momento Client SDKs Specification: Client Configuration

This document provides information on how Momento client implementations must expose configuration options to users.

All code in this example repo is TypeScript pseudocode.  Implementations should use the appropriate idioms for the
programming language at hand.

# Tenets

The overarching goals of the specification are:

> * Configuration should be as simple as possible for users.  We do testing and tuning work up front to make sure

Reviewer: ```suggestion
* Configuration should be as simple as possible for users.  We do testing and tuning work up front to make sure we
```

# Choosing a configuration

The `SimpleCacheClient` constructor should accept a required argument whose type is our `Configuration` interface.  e.g.:

```typescript
import * as config from '@gomomento/sdk/config';

const client = SimpleCacheClient({config: config.Dev.latest()});
```

# Pre-built Configuration Objects

## Requirements for each pre-built configuration objects

Each SDK must provide a namespace (via the idiomatic construct in the relevant programming language) that contains
an enumerated list of pre-built configuration objects.

All configuration objects must provide a `latest()` function that returns the latest "blessed" defaults from Momento.
Documentation must clearly indicate that these values may change over time if Momento finds in the future that some
different default values will provide a better or more performant user experience.

All configuration objects must also provide a `default()` function that returns the original "blessed" defaults from
Momento.  Documentation must clearly state that these values will not change over time; users can rely on them to
be stable across new minor version releases of the SDKs.

> The `SimpleCacheClient` constructor should expose a required argument that is an instance of

Reviewer: instance of a config?

Author: fixed

## Required pre-built configuration objects

Each SDK must provide at least the following out-of-the-box configuration objects:

```typescript
import * as config from '@gomomento/sdk/config';

/**
 * Dev provides defaults suitable for a medium-to-high-latency dev environment.  Permissive timeouts, retries, potentially
 * a higher number of connections, etc.
 */
const devConfig = config.Dev.latest();

/**
 * ProdLowLatency provides defaults suitable for a low-latency prod environment.  This is for the use case where the
 * client is running on an application server in the same region as the Momento service.  It has more agressive timeouts
 * and retry behavior, to ensure that cache unavailability doesn't force unacceptably high application latencies.
 */
const prodConfig = config.ProdLowLatency.latest();
```

## Configuration interface

The configuration interface provides nested objects for retry strategy, middleware, and transport:

```typescript
interface Configuration {
  retryStrategy: RetryStrategy,

>   middleware: Middleware,

Reviewer: I feel like it might be more idiomatic to have this be `middlewares`, which would take in an array of middleware. I know that this impl would support multiple middlewares, its just that the impl of supporting those middlewares would be the customers job instead of ours.

Author: Yeah fair callout, I'll think about this.  We could certainly provide a CompositeMiddleware class so they wouldn't have to invent it.  But in this case I'm on the fence, you might be right.  Do you have any particular software in mind that you are using as your mental foundation for what is "idiomatic" for middleware?

Reviewer: I think the one that comes to mind first is express js. The way its normally implemented is
```
app.use(cookieParser())
app.use(auth())
etc...
```

Reviewer: I am also on the fence about it tho, just thought it was worth mentioning

Author: changed to an array

  transportStrategy: TransportStrategy
}
```

### RetryStrategy

The `RetryStrategy` provides a function that is called when a request fails.  This function is responsible for determining
whether the request should be retried, and if so, how long the client should delay before retrying:

```typescript
interface RetryStrategy {
  /**
   * Returns the number of milliseconds after which the request should be retried, or `undefined` if the request should
   * not be retried.  `0` means retry immediately.
   */

>   determineWhenToRetryRequest(grpcResult: GrpcResult, grpcRequest: GrpcRequest, attemptNumber: number): Promise<number | undefined>;

Reviewer 2: If we were to support http as transport down line would we just add another field on this interface with different signature like this? 
```typescript
   determineWhenToRetryRequest(httpResult: httpResult, HttpRequest: HttpRequest, attemptNumber: number): Promise<number | undefined>;
```

Just want to make sure im understanding how your thinking about modifying this down line with different transport strategies. Just warning flag went off when I see GRPC specific things in method signature exposed to user.

Author: Yeah, again - this would never be visible to a user that chose to use one of our out-of-the-box configurations.  But it would be visible if they wanted to implement a custom configuration.

If they just want to tune a "max retries" setting or something, we may have a builder like:

```typescript
const retryStrategy = new BasicRetryStrategy({
   maxAttempts: 5,
   exponentialBackOffFactor: 2,
   jitterNanos: 500
});
```

That thing would implement the interface, and then they'd plug that in to their Configuration constructor.  So they will only ever see the gRPC stuff in the signature if they are doing a custom implementation of a RetryStrategy, which is most likely an indicator that we should have provided more/better out-of-the-box ones.

But yes, I also think/hope that the interface could be extended to support alternate signatures for HTTP in the future.

Thanks for reviewing this thoroughly and challenging.  I'll add more example code to cover this case as well, but please continue to raise questions or push back if you still have concerns!

Author: added some examples in https://github.com/momentohq/standards-and-practices/pull/14/commits/7c54ee1016b39bef24280ead497e121de398e974

}
```

The simplest default implementation of this strategy can expose a `maxAttempts` integer, and always return `0` for
immediate retries until the maximum number of attempts is reached.  A slightly more sophisticated initial implementation
would have basic support for exponential backoff and jitter.

### Middleware

The `Middleware` interface allows the Configuration to provide a higher-order function that wraps all requests.  This
allows future support for things like client-side metrics or other diagnostics helpers.

```typescript
interface Middleware {

>   wrapRequest(middlewareFn: (GrpcRequest) => GrpcResult): GrpcResult;

Reviewer 2: Should the signature for our middleware be protocol agnostic? Should we be able to switch out GRPC without the user's middleware breaking? 

Just want to call out since before we had been trying to hide GRPC from any client-facing exposed bits in past.

Author: This is a great question.

I would expect that most users will never implement this, they will just use our pre-builts.

For the advanced users that do want to implement this, I... think it's probably fair game to say that they have opted in to that complexity?  For the middleware to be maximally useful I think we would have to provide a wrapper data class that included all of the gRPC info (metadata etc.) anyway, so I'm not sure it's worth the abstraction?  But I could be convinced to reconsider.

Reviewer 2: I think this is ok and safest place to start. We can always revisit in v2 version if we get alot of feedback. Would rather avoid abstracting to much upfront.

Author: agree

users as much as possible, but we will need to provide different gRPC configurations for dev vs. prod environments.
And ultimately we will need to support advanced customers who need to mess with these knobs.

```typescript
interface TransportStrategy {
  grpcConfig: GrpcConfiguration
}

interface GrpcConfiguration {
  numChannels(): number

  // These interface functions expose individual settings for gRPC channels.  They are just here to ensure that
  // strategy implementations provide values for settings that we know to be important.  These may vary by language
  // since the gRPC implementations in each language have subtly different behaviors.
  maxSessionMemory(): number
  useLocalSubChannelPool(): boolean

  // This is a map that encapsulates the settings above, and may also include other channel-specific settings.
  // This allows strategy implementations to provide gRPC config key/value pairs for any available setting, even
  // if it's not one we've explicitly tried / recommended.  The strategy implementation should implement this by
  // calling the functions above, along with allowing a mechanism for specifying additional key/value pairs.

>   grpcChannelConfig(): Record<string, string>

Reviewer 2: so as end-user would I pass a  config like this:
```typescript
const client = new SimpleCacheClient({
  config: {
    numChannels: 4,
    grpcChannelConfig: {
      'numChannels': '6'
      'ignoreUnknownServices': 'true'
    }
  }
})
```

Obviously this is weird use case passing `numChannels` twice but would this work from end user perspective? They would need to pass each overide as strings? Would overides in `grpcChannelConfig` take precedent higher up fields? 

Maybe just add a small section on using this for advanced usage overrides with an example or 2?

Author: Yeah that's not exactly what I have in mind, but I can provide an advanced usage example, that's a fair request.

Author: added some examples in https://github.com/momentohq/standards-and-practices/pull/14/commits/7c54ee1016b39bef24280ead497e121de398e974

<img src="https://docs.momentohq.com/img/logo.svg" alt="logo" width="400"/>

# Momento Client SDKs Specification: Client Configuration

This document provides information on how Momento client implementations must expose configuration options to users.

All code in this example repo is TypeScript pseudocode.  Implementations should use the appropriate idioms for the
programming language at hand.

# Tenets

The overarching goals of the specification are:

* Configuration should be as simple as possible for users.  We do testing and tuning work up front to make sure

>   have established high-confidence default settings that cover the 90% use case, so that users don't have to spend

Reviewer 2: ```suggestion
we have established high-confidence default settings that cover the 90% use case, so that users don't have to spend
```

  time figuring it out themselves.

> * Because there are different requirements for different environments (e.g. dev vs. prod), we provide a small handful of

Reviewer 2: Maybe switch to `laptop vs cloud or in-region` per other convos we have had?

> * Because there are different requirements for different environments (e.g. dev vs. prod), we provide a small handful of

Author: ```suggestion
* Because there are different requirements for different environments (e.g. laptop vs. in-region), we provide a small handful of
```

Reviewer 2: LGTM

  pre-built configurations that users can choose from, out-of-the box.
* Configurations must be specified as interfaces, so that alternate implementations can be added over time.
* Users should always have a last-resort option of manually tuning the settings themselves, in case there are special
  cases that our defaults are not quite right for.

# Choosing a configuration

The `SimpleCacheClient` constructor should accept a required argument whose type is our `Configuration` interface.  e.g.:

```typescript
import * as config from '@gomomento/sdk/config';

> const client = SimpleCacheClient({config: config.Laptop.latest()});

Reviewer 2: I like the config.Laptop seeing it used here now. I think is better then `Dev` for sure

```

# Pre-built Configuration Objects

## Requirements for each pre-built configuration objects

Each SDK must provide a namespace (via the idiomatic construct in the relevant programming language) that contains
an enumerated list of pre-built configuration objects.

> All configuration objects must provide a `latest()` function that returns the latest "blessed" defaults from Momento.

Reviewer 2: Just curious what is alternative to `latest()` could I pin to older version of config down the line? Is it possible to include example how that might look?

Author: see: https://www.notion.so/momentohq/Design-Proposal-SDK-Configuration-UX-b7086fa5d1e34f629904c14477a87554#6e5c5229d2844d578214331848584e6f

For now I am only spec'ing `latest()`, to give us maximum flexibility for adding in the stable ones later.

  // calling the functions above, along with allowing a mechanism for specifying additional key/value pairs.
  grpcChannelConfig(): Record<string, string>;
}
```

# Advanced Configuration Examples

In this section we expand a bit on how a customer might define their own advanced configuration if they want to
opt out of our pre-builts.

Here is a sketch of what the implementation of one of the pre-built configurations might look like inside of the
SDK (users can see the source code for this if they like):

```typescript
namespace InRegion {
  export class LowLatency {
    static latest(): Configuration {
      return {
        retryStrategy: new ExponentialBackoffStrategy({
          maxAttempts: 3,
          backoffMilliseconds: 0.5,
          jitterMilliseconds: 0.5,
          exponentialBackoffFactor: 2,
        }),
        middlewares: [], // TODO: this could eventually include some basic client-side metrics OOTB
        transportStrategy: {
          maxConcurrentRequests: 2,
          grpcConfig: new BasicGrpcConfiguration({
            numChannels: 5
          })
        }
      }
    }
  }

> }

Reviewer 2: 🧼 Clean!

```

If a customer wishes to override the defaults, they can do something like this:

```typescript
const myConfig = configs.InRegion.LowLatency.latest()
  .withTransportStrategy({
    maxConcurrentRequests: 2,
    grpcConfig: new BasicGrpcConfiguration({
      numChannels: 5,
      additionalGrpcChannelSettings: {

>               'foo': 'bar'

Reviewer 2: is this only string settings? If can take numbers maybe pass example of overiding that also. If not I think is fine as is.

Author: i'm pretty sure it's string->string.
